### PR TITLE
Add explicit quota support to CTFE

### PR DESCRIPTION
**Pull request by AlCutter**
This PR updates CTFE to use the new Trillian log API support for explicit request quotas (added in [Trillian #1143](https://github.com/google/trillian/pull/1143)).
CTFE will add `ChargeTo` entries for the IP making the incoming request, as well as entries for each intermediate certificate in submitted chains.

The new quota requests may be disabled with flags defined in `trillian/ctfe/ct_server/main.go`

Should help with #244 .

### CODE REVIEW

File: trillian/ctfe/cert_quota.go [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+// with intermediate certificates.
+const CertificateQuotaUserPrefix = "@intermediate"
+
+// QuotaUserForCert returns a User quota id string for the passed in
+// certificate.
+// This is intended to be used for quota limiting by intermediate certificates,
+// but the function does not enforce anything about the passed in cert.
+//
+// Format returned is:
+//   "CertificateQuotaUserPrefix Subject SHA256_hex(SubjectPublicKeyInfo[0:5])"
```
Comment (daviddrysdale) [inline trillian/ctfe/cert_quota.go]: `SHA256_hex(SubjectPublicKeyInfo)[0:5]`  I think (different end bracket order).

File: trillian/ctfe/cert_quota_test.go [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+			cert: mustDePEM(t, testonly.CACertPEM),
+			want: "@intermediate O=Certificate Transparency CA,L=Erw Wen,ST=Wales,C=GB 02adddca08",
+		},
+		{
+			desc: "intermediate",
+			cert: mustDePEM(t, testonly.FakeIntermediateCertPEM),
+			want: "@intermediate CN=FakeIntermediateAuthority,OU=Eng,O=Google,L=London,ST=London,C=GB 6e62e56f67",
+		},
+	} {
+		if got := QuotaUserForCert(test.cert); got != test.want {
```
Comment (daviddrysdale) [inline trillian/ctfe/cert_quota_test.go]: Be nice to get into the habit of using
```go
t.Run(test.desc, func(t *testing.T) {
...
```
style.

File: trillian/ctfe/cert_quota_test.go [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+			t.Errorf("%s: got %q, want %q", test.desc, got, test.want)
```
Comment (daviddrysdale) [inline trillian/ctfe/cert_quota_test.go]: Canonical form would be more like `t.Errorf("QuotaUserForCert()=%q, want %q", got, test.want)`

File: trillian/ctfe/instance.go [hunk 4]
```diff
@@ -93,6 +94,17 @@ type InstanceOptions struct {
 	// a boolean to indicate whether the conversion succeeded.
 	ErrorMapper func(error) (int, bool)
 	RequestLog  RequestLog
+	// RemoteUser returns a string representing the originating host for the
+	// given request. This string will be used as a User quota key.
+	// If unset, no quota will be requested for remote users.
+	RemoteQuotaUser func(*http.Request) string
+
+	// CertificateQuotaUser returns a string represeing the passed in
```
Comment (daviddrysdale) [inline trillian/ctfe/instance.go]: "representing"

File: trillian/ctfe/instance.go [hunk 5]
```diff
@@ -93,6 +94,17 @@ type InstanceOptions struct {
...
+	// intermediate certificate. This string will be user as a User quota key for
+	// the cert.  Quota will be requested for each intermediate in an add chain
```
Comment (daviddrysdale) [inline trillian/ctfe/instance.go]: "in an add-[pre-]chain..."

File: trillian/ctfe/handlers_test.go [hunk 6]
```diff
@@ -108,6 +109,41 @@ type handlerTestInfo struct {
 	c             *LogContext
 }
 
+const certQuotaPrefix = "CERT:"
+
+func quotaUserForCert(c *x509.Certificate) string {
+	return fmt.Sprintf("%s %s", certQuotaPrefix, c.Subject.String())
+}
+
+func quotaUsersForChain(t *testing.T, pem ...string) []string {
```
Comment (daviddrysdale) [inline trillian/ctfe/handlers_test.go]: There's a slight chance of confusion here because the "chain" is different from what the mainline code calls a "chain" -- the main code is `[1:]` but this is `[:]`.

File: trillian/ctfe/instance.go [hunk 7]
```diff
@@ -93,6 +94,17 @@ type InstanceOptions struct {
 	// a boolean to indicate whether the conversion succeeded.
 	ErrorMapper func(error) (int, bool)
 	RequestLog  RequestLog
+	// RemoteUser returns a string representing the originating host for the
+	// given request. This string will be used as a User quota key.
```
Comment (daviddrysdale) [inline trillian/ctfe/instance.go]: (assuming per-log config)
"will be used as a User quota key for any log that has per-user charging configured"

File: trillian/ctfe/handlers_test.go [hunk 8]
```diff
@@ -366,7 +446,14 @@ func TestAddChain(t *testing.T) {
 				}
 			}
 			rsp := trillian.QueueLeavesResponse{QueuedLeaves: queuedLeaves}
-			info.client.EXPECT().QueueLeaves(deadlineMatcher(), &trillian.QueueLeavesRequest{LogId: 0x42, Leaves: leaves}).Return(&rsp, test.err)
+			req := &trillian.QueueLeavesRequest{LogId: 0x42, Leaves: leaves}
+			if len(test.wantQuotaUsers) > 0 {
+				req.ChargeTo = &trillian.ChargeTo{}
```
Comment (daviddrysdale) [inline trillian/ctfe/handlers_test.go]: why not just
```go
req.ChargeTo = &trillian.ChargeTo{User: test.wantQuotaUsers}
```

File: trillian/ctfe/handlers_test.go [hunk 9]
```diff
@@ -340,6 +413,13 @@ func TestAddChain(t *testing.T) {
 	defer info.mockCtrl.Finish()
 
 	for _, test := range tests {
+		if test.enableRemoteQuota {
+			// first entry in wantQuotaUsers is the remote user
+			info.setRemoteQuotaUser(test.wantQuotaUsers[0])
```
Comment (daviddrysdale) [inline trillian/ctfe/handlers_test.go]: This is a bit hard to follow (pulling values from the expected test output to serve as injected inputs) -- maybe just use a hard-coded remote-user string and shift `setRemoteQuotaUser()` to take a boolean?

File: trillian/ctfe/ct_server/main.go [hunk 10]
```diff
@@ -61,6 +62,8 @@ var (
 	tracing            = flag.Bool("tracing", false, "If true opencensus Stackdriver tracing will be enabled. See https://opencensus.io/.")
 	tracingProjectID   = flag.String("tracing_project_id", "", "project ID to pass to stackdriver. Can be empty for GCP, consult docs for other platforms.")
 	tracingPercent     = flag.Int("tracing_percent", 0, "Percent of requests to be traced. Zero is a special case to use the DefaultSampler")
+	quotaRemote        = flag.Bool("quota_remote", true, "Enable requesting of quota for IP address sending incoming requests")
```
[the remark this answers is not in the document]
Comment (AlCutter) [reply trillian/ctfe/ct_server/main.go]: I'm going to call YAGNI and punt that for now, happy to implement later if we determine a real need for it.

File: trillian/ctfe/ct_server/main.go [hunk 11]
```diff
@@ -61,6 +62,8 @@ var (
...
+	quotaIntermediate  = flag.Bool("quota_intermediate", true, "Enable requestion of quota for intermediate certificates in sumbmitted chains")
```
[the remark this answers is not in the document]
Comment (AlCutter) [reply trillian/ctfe/ct_server/main.go]: It's like requesting but bettered.
fixed.

File: trillian/ctfe/cert_quota.go (hunk 1 above)
Comment (AlCutter) [reply trillian/ctfe/cert_quota.go]: They're both wrong :)
Yours 'cos that'll be 5 ASCII bytes of hex with my imaginary, and mine because it'll be the wrong hash...
fixed by separating the sha256 and hexification

File: trillian/ctfe/cert_quota_test.go (hunk 2 above)
Comment (AlCutter) [reply trillian/ctfe/cert_quota_test.go]: Seems overkill for this (+2 lines of delta for no real benefit other than habit formint), but done anyway.

File: trillian/ctfe/cert_quota_test.go (hunk 3 above)
Comment (AlCutter) [reply trillian/ctfe/cert_quota_test.go]: done

File: trillian/ctfe/instance.go (hunk 5 above)
Comment (AlCutter) [reply trillian/ctfe/instance.go]: done

File: trillian/ctfe/handlers_test.go (hunk 6 above)
Comment (AlCutter) [reply trillian/ctfe/handlers_test.go]: Good call, changed to `quotaUsersForIssuers`.

File: trillian/ctfe/instance.go (hunk 7 above)
Comment (AlCutter) [reply trillian/ctfe/instance.go]: Punting that for now

File: trillian/ctfe/handlers_test.go (hunk 8 above)
Comment (AlCutter) [reply trillian/ctfe/handlers_test.go]: Thanks!

File: trillian/ctfe/handlers_test.go (hunk 9 above)
Comment (AlCutter) [reply trillian/ctfe/handlers_test.go]: changed

**Comment by AlCutter** [comment]
PTAL, ta!

File: trillian/ctfe/handlers.go [hunk 12]
```diff
@@ -320,6 +320,18 @@ func parseBodyAsJSONChain(c *LogContext, r *http.Request) (ct.AddChainRequest, e
 	return req, nil
 }
 
+// appendUserCharge adds the specified user to the passed in ChargeTo and
+// and returns the result.
+// If the passed-in ChargeTo is nil, then a new one is created with the passed
+// in user and returned.
+func appendUserCharge(a *trillian.ChargeTo, user string) *trillian.ChargeTo {
```
Comment (daviddrysdale) [inline trillian/ctfe/handlers.go]: Grr, Github ate a large comment of mine -- let me try to recreate it.

I think things could be simpler if there was also a `func (c *LogContext) chargeUser(r *http.Request) *trillian.ChargeTo` function which could be used in-line in request structure initialization in most places.  The only special case would then be the addChainInternal case:
```go
	req := trillian.QueueLeavesRequest{
		LogId: c.logID,
		Leaves: []*trillian.LogLeaf{&leaf},
		ChargeTo: c.chargeUser(r),
	}
	if c.instanceOpts.CertificateQuotaUser != nil {
		// TODO(al): ignore pre-issuers? Probably doesn't matter
		for _, cert := range chain[1:] {
			req.ChargeTo = appendUserCharge(req.ChargeTo, c.instanceOpts.CertificateQuotaUser(cert))
		}
	}	
```
